Cross-Platform Geb FileInput module support

matrei opened the pull request:
This resolves an issue on Windows when using the file from `container.copyFileToContainer()` with the Geb FileInput module.

As this is a new feature I created it against a new 4.2.x branch for Grails 6.

matrei committed:
feat: cross-platform Geb FileInput module support

This resolves an issue on Windows when using the file from `container.copyFileToContainer()` with the Geb FileInput module.

matrei committed:
ci: fix error in os matrix

matrei committed:
chore: go to 4.2.0-SNAPSHOT

matrei committed:
ci: remove `macos-latest` from the os matrix

`macos-latest` does not seem to have container support.

matrei committed:
ci: remove `windows-latest` from the os matrix

`windows-latest` does not seem to provide a container environment.

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -10,7 +10,10 @@ on:
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [ubuntu-24.04] # windows-latest and macos-latest do not provide a container environment
```
https://github.com/actions/runner/issues/904

Looks like there may be a work around to use docker ce.  Maybe create a separate build so we can have test coverage?

matrei replied:
I could find this workaround in that issue discussion, where did you read about it.

This PR also has the issue that it presupposes that containers are Linux based.
But, right now, it seems Testcontainers does not support Windows containers on Windows anyways: https://java.testcontainers.org/supported_docker_environment/windows/#windows-container-on-windows-wcow

jdaugherty replied:
I think the work around was this comment: 
https://github.com/actions/runner/issues/904#issuecomment-855230463

They use docker ce instead of the enterprise docker.  

As for test containers, isn't the issue running tests on windows executing in a linux container? I don't think the code you've written will execute in the container.  Test containers definitely works when running on windows and the container is inside of docker / a linux vm.

matrei replied:
I don't know, we could give it a try I guess. The comment is from 2021.

> isn't the issue running tests on windows executing in a linux container

Yes, Windows Github runner (e.g. host) running selenium in Linux (e.g. container). 
But if you run Windows host and Windows container, will it work then? Maybe, it could work with forward slashes in windows.

jdaugherty replied:
I don't think we'll ever run in a windows container - there isn't  desktop environment in windows containers.  I think we'll always end up targeting windows (test location) and linux container (browser location).  

I dug into this a little more and https://github.com/orgs/community/discussions/29255#discussioncomment-4020736 explains why this isn't possible.  We can't have nested virtualization.  We'll have to settle for running on ubuntu and not having test coverage for the windows scenarios =(

jdaugherty commented on src/testFixtures/groovy/grails/plugin/geb/support/ContainerGebFileInputSource.groovy:
```diff
@@ -0,0 +1,50 @@
+     */
+    protected String normalizePath(String path) {
+        // Replace backslashes with forward slashes for Windows compatibility
+        path.replaceAll('\\\\', '/')
```
If you use 
    
     path.replace(File. pathSeparatorChar, '/' as char)

Wouldn't this make it so the file separator is only replaced on windows?  It also would avoid the pattern compilation.  Technically the \\ is valid on other operating systems ...

matrei replied:
Good points! But should it not use `File.separatorChar`?

jdaugherty replied:
Yes!

matrei committed:
refactor: optimize path normalization

- Use `replace(char, char)` instead of `replaceAll(String, String)` to avoid the overhead of regex pattern compilation.
- Replace hardcoded string literals with `File.separatorChar`, ensuring the `replace` operation is a no-op on non-Windows hosts where no substitution is needed.
